Kernel: Add an SD card driver for the Raspberry Pi port

mrkct opened the pull request:
Pinging @Ollrogge because we worked on this together, and @FireFox317 because we used part of his code to start on this.

These commits add a simple SD card driver for the raspberry pi port.
This brings us further in the boot process, we now load SystemServer and crash because `fork` is not implemented yet.

It also removes the ramdisk from the aarch64 build since it's no longer needed. We can now build aarch64 again by simply running  "Meta/serenity.sh run aarch64".

Also an unrelated commit in Jail.h since the missing include broke the aarch64 build.

drunderscore commented:
These commits seem out of order... we should add the SD card driver, then make aarch64 run with the disk image in the SD card slot, and then disable the ramdisk!

Hendiadyoin1 commented on Kernel/Storage/SD/Commands.h:
```diff
@@ -0,0 +1,156 @@
+    u8 direction : 1;
+    u8 multiblock : 1;
+    u16 resp_b : 10;
+    u8 response_type : 2;
+    u8 res0 : 1;
+    u8 crc_enable : 1;
+    u8 idx_enable : 1;
+    u8 is_data : 1;
+    u8 type : 2;
+    u8 index : 6;
```
```suggestion
    ResponseType response_type : 2;
    u8 res0 : 1;
    u8 crc_enable : 1;
    u8 idx_enable : 1;
    u8 is_data : 1;
    u8 type : 2;
    CommandIndex index : 6;
    ```
    This makes the successors way simpler
    (which could also have been a static_cast, which might be checked by UB-san)

mrkct replied:
Neat, didn't know we could do that

Hendiadyoin1 commented on Kernel/Storage/SD/Registers.h:
```diff
@@ -0,0 +1,226 @@
+    u32 shared_bus_control;
+    u32 reserved_1[6];
+    u32 slot_interrupt_status_and_version;
+} __attribute__((packed));
```
prefer `[[gnu::packed]]` next to the Struct name, like in the rest of the code base

Hendiadyoin1 commented on Kernel/Storage/SD/Registers.h:
```diff
@@ -0,0 +1,226 @@
+    u32 slot_interrupt_status_and_version;
+} __attribute__((packed));
+
+struct OperatingConditionRegister {
+    u32 : 15;
+    u32 vdd_voltage_window_27_28 : 1;
+    u32 vdd_voltage_window_28_29 : 1;
+    u32 vdd_voltage_window_29_30 : 1;
+    u32 vdd_voltage_window_30_31 : 1;
+    u32 vdd_voltage_window_31_32 : 1;
+    u32 vdd_voltage_window_32_33 : 1;
+    u32 vdd_voltage_window_33_34 : 1;
+    u32 vdd_voltage_window_34_35 : 1;
+    u32 vdd_voltage_window_35_36 : 1;
+    u32 switching_to_18v_accepted : 1;
+    u32 : 2;
+    u32 over_2tb_support_status : 1;
+    u32 : 1;
+    u32 uhs2_card_status : 1;
+    u32 card_capacity_status : 1;
+    u32 card_power_up_status : 1;
+
+    static OperatingConditionRegister from_acmd41_response(u32 value)
+    {
+        union {
+            u32 x;
+            struct OperatingConditionRegister ocr;
+        } u;
+        u.x = value;
+        return u.ocr;
+    }
+} __attribute__((packed));
```
See above, and not necessary here anyway

Hendiadyoin1 replied:
Why not make this a full union with a `raw` u32 member and an anonymous struct holding all others?

Hendiadyoin1 commented on Kernel/Storage/SD/Registers.h:
```diff
@@ -0,0 +1,226 @@
+
+// SD Host Controller Simplified Specification Version 3.00
+// NOTE: The registers must be 32 bits, because of a quirk in the RPI.
+struct SDRegisters {
```
This seems to be missing a size assertion

nico replied:
(look for `static_assert(AssertSize<` for many examples)

Hendiadyoin1 commented on Kernel/Storage/SD/Registers.h:
```diff
@@ -0,0 +1,226 @@
+        return u.ocr;
+    }
+} __attribute__((packed));
+static_assert(sizeof(OperatingConditionRegister) == sizeof(u32));
```
prefer `static_assert(AssertSize<Type, Size>())`

Hendiadyoin1 commented on Kernel/Storage/SD/Commands.h:
```diff
@@ -0,0 +1,156 @@
+
+namespace Kernel::SD {
+
+// Commands, defined in PLSS 4.7.4 with the format described in BCM2835 "CMDTM
+// Register"
```
bogus line-break

Hendiadyoin1 commented on Kernel/Storage/SD/SDHostController.cpp:
```diff
@@ -0,0 +1,671 @@
+        ;
+}
+
+constexpr u32 MAX_SUPPORTED_SDSC_FREQUENCY = 25000000;
+
+// In "m_registers->host_configuration_1"
+// In sub-register "Clock Control"
+constexpr u32 INTERNAL_CLOCK_ENABLE = 1 << 0;
+constexpr u32 INTERNAL_CLOCK_STABLE = 1 << 1;
+constexpr u32 SD_CLOCK_ENABLE = 1 << 2;
+
+// In sub-register "Software Reset"
+constexpr u32 SOFTWARE_RESET_FOR_ALL = 0x01000000;
+
+// In Interrupt Status Register
+const u32 COMMAND_COMPLETE = 1 << 0;
+const u32 TRANSFER_COMPLETE = 1 << 1;
+const u32 BUFFER_WRITE_READY = 1 << 4;
+const u32 BUFFER_READ_READY = 1 << 5;
+
+// PLSS 5.1: all voltage windows
+constexpr u32 ACMD41_VOLTAGE = 0x00ff8000;
+// PLSS 4.2.3.1: All voltage windows, XPC = 1, SDHC = 1
+constexpr u32 ACMD41_ARG = 0x50ff8000;
+
+constexpr u32 BLOCK_LEN = 512;
```
Wither defines and UPPERCASE or `constexpr` and snake_case

Hendiadyoin1 commented on Kernel/Storage/SD/SDHostController.cpp:
```diff
@@ -0,0 +1,671 @@
+    m_registers->interrupt_status = COMMAND_COMPLETE;
+
+    // 3. Read the Response register(s) to get the response.
+    struct Response r = {};
```
no need for the `struct` specifier
(I think you used that in other places as well)

Hendiadyoin1 commented on Kernel/Storage/SD/SDHostController.cpp:
```diff
@@ -0,0 +1,671 @@
+
+ErrorOr<u32> SDHostController::retrieve_sd_clock_frequency()
+{
+    const i64 ONE_MHZ = 1'000'000;
+    const u32 bclock = ((m_registers->capabilities_0 & 0xff00) >> 8) * ONE_MHZ;
```
why `const`?

Hendiadyoin1 commented on Kernel/Storage/SD/SDHostController.cpp:
```diff
@@ -0,0 +1,671 @@
+
+ErrorOr<u32> SDHostController::retrieve_sd_clock_frequency()
+{
+    const i64 ONE_MHZ = 1'000'000;
```
see above

Hendiadyoin1 commented on Kernel/Storage/SD/SDHostController.h:
```diff
@@ -0,0 +1,112 @@
+        return m_registers->present_state & CARD_INSERTED;
+    }
+
+    enum class SDHostVersion {
+        Version1 = 0x0,
+        Version2 = 0x1,
+        Version3 = 0x2,
+        Unknown
+    };
+
+    SDHostVersion host_version()
+    {
+        const u16 host_controller_version_register = m_registers->slot_interrupt_status_and_version >> 16;
+        switch (host_controller_version_register & 0xff) {
+        case 0x0:
+            return SDHostVersion::Version1;
+        case 0x1:
+            return SDHostVersion::Version2;
+        case 0x2:
+            return SDHostVersion::Version3;
+        default:
+            return SDHostVersion::Unknown;
+        }
+    }
```
You can make this member an enum and bitfiled to simplify this

Hendiadyoin1 commented on Kernel/Storage/SD/SDHostController.h:
```diff
@@ -0,0 +1,112 @@
+        Read,
+        Write
+    };
+    ErrorOr<void> transaction_control_with_data_transfer_using_the_dat_line_without_dma(
+        SD::CommandIndex,
+        u32 argument,
+        u32 block_count,
+        u32 block_size,
+        UserOrKernelBuffer,
+        DataTransferType data_transfer_type);
```
one line please

Hendiadyoin1 commented on Kernel/Storage/SD/SDHostController.h:
```diff
@@ -0,0 +1,112 @@
+
+    bool is_card_inserted() const
+    {
+        constexpr u32 CARD_INSERTED = 1 << 16;
```
see above

Hendiadyoin1 commented on Kernel/Storage/SD/SDMemoryCard.cpp:
```diff
@@ -0,0 +1,60 @@
+
+namespace Kernel {
+
+SDMemoryCard::SDMemoryCard(SDHostController& sdhc,
+    StorageDevice::LUNAddress lun_address,
+    u32 hardware_relative_controller_id, u32 block_len,
+    u64 capacity_in_blocks, u32 relative_card_address,
+    SD::OperatingConditionRegister ocr,
+    SD::CardIdentificationRegister cid,
+    SD::SDConfigurationRegister scr)
```
one line

Hendiadyoin1 commented on Kernel/Storage/SD/SDMemoryCard.h:
```diff
@@ -0,0 +1,58 @@
+    virtual void start_request(AsyncBlockDeviceRequest&) override;
+
+private:
+    SDMemoryCard(
+        SDHostController& sdhc,
+        StorageDevice::LUNAddress,
+        u32 hardware_relative_controller_id,
+        u32 block_len,
+        u64 capacity_in_blocks,
+        u32 relative_card_address,
+        SD::OperatingConditionRegister ocr,
+        SD::CardIdentificationRegister cid,
+        SD::SDConfigurationRegister scr);
```
see above

Hendiadyoin1 commented on Kernel/Storage/SD/Commands.h:
maybe designate underlying type to the enums

Hendiadyoin1 reviewed (commented):
Nice work,

A lot of style notes from my side

nico commented on Kernel/Arch/aarch64/RPi/GPIO.cpp:
```diff
@@ -94,4 +94,26 @@ void GPIO::internal_enable_pins(u32 enable[2], PullUpDownState state)
 
+void GPIO::set_pin_high_detect_enable(unsigned pin_number, bool enable)
+{
+    u32 data[2] = {};
```
What's the advantage of loading both words into a local array? Doesn't

```
    if (enable) {
        if (pin_number < 32)
            m_registers->high_detect_enable.bits[0] |= (1 << pin_number);
        else
            m_registers->high_detect_enable.bits[1] |= (1 << (pin_number - 32));
    } else {
        if (pin_number < 32)
            m_registers->high_detect_enable.bits[0] &= ~(1 << pin_number);
        else
            m_registers->high_detect_enable.bits[1] &= ~(1 << (pin_number - 32));
    }
```

do the same thing with less code?

mrkct replied:
No advantage, this was code from [FireFox317](https://github.com/FireFox317) I mindlessly copy-pasted.

I've changed it to yours.

nico commented on Kernel/Arch/aarch64/RPi/SDHostController.cpp:
```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright (c) 2023, the SerenityOS developers.
```
feel free to add yourself up here if you want!

Ollrogge replied:
Nah all good :) I just helped

nico reviewed (commented):
Very cool. I only glanced.

Once this is in, we can revert 187723776a1140085a187b80effb82f31cc6bb9b I think (?)

supercomputer7 commented:
> Very cool. I only glanced.
> 
> Once this is in, we can revert [1877237](https://github.com/SerenityOS/serenity/commit/187723776a1140085a187b80effb82f31cc6bb9b) I think (?)

We should revert it (the commit which reverts dropping ramdisk support).
As I said before, `initramfs` is the way to go and it will be introduced when I have time to rebuild the prekernel stage but if we don't need ramdisks for aarch64, it should be removed immediately.

Panky-codes commented on Kernel/Storage/SD/SDHostController.h:
```diff
@@ -0,0 +1,86 @@
+
+private:
+    ErrorOr<void> initialize();
+    ErrorOr<NonnullLockRefPtr<SDMemoryCard>> try_initialize_inserted_card();
```
Do we need the SDMemoryCard to be of type `NonNullLockRefPtr` instead of a plain `NonNullRefPtr`? From what I see, after Init, you don't touch the m_card at all.
I think there is a general effort to reduce the usage of `NonNullLockRefPtr`.

mrkct replied:
Right now `SDMemoryCard` does not need to be in a `LockRefPtr`, but there's an overridden method `device` in `SDHostController` that must return a `LockRefPtr`.

Is it safe to store `SDMemoryCard` in a normal `RefPtr` and then have `device` look something like this, which converts from `SDMemoryCard*` to `LockRefPtr<SDMemoryCard>`?

```cpp
virtual LockRefPtr<StorageDevice> device(u32 index) const override { return index == 0 ? m_card.ptr() : nullptr; }
```

Panky-codes replied:
You are right. I think you can leave it as is. We probably need to change how we manage `StorageDevice` in Storagemanagement, then we can change this to return RefPtr instead of LockRefPtr.

Hendiadyoin1 commented on Kernel/Storage/SD/Commands.cpp:
Do you maybe have spec links for the commands?

mrkct replied:
Added links to the page with all the specs instead of direct ones because there's a wall of text to accept before they let you download them: https://www.sdcard.org/downloads/pls/

Hendiadyoin1 commented on Kernel/Storage/SD/Commands.h:
```diff
@@ -0,0 +1,88 @@
+    AppCmd = 55,
+};
+
+enum class CommandType : u8 { Normal,
```
`Normal`, should get it's own line

Hendiadyoin1 replied:
Same in the enums below

Hendiadyoin1 commented on Kernel/Storage/SD/Commands.h:
```diff
@@ -0,0 +1,88 @@
+union Command {
+    u32 raw;
+
+    struct [[gnu::packed]] {
+        u8 : 1;
+        bool block_counter : 1;
+        SendAutoCommand auto_command : 2;
+        DataTransferDirection direction : 1;
+        bool multiblock : 1;
+        u16 : 10;
+        ResponseType response_type : 2;
+        u8 : 1;
+        bool crc_enable : 1;
+        bool idx_enable : 1;
+        bool is_data : 1;
+        CommandType type : 2;
+        CommandIndex index : 6;
+        u8 : 2;
+    };
```
prefer to use one type for the bit field type, when its not an enum
like `u32` in the case where the final size is 32 bits

Hendiadyoin1 replied:
Also I think this does not need to be explicitely packed

Hendiadyoin1 commented on Kernel/Storage/SD/Registers.h:
```diff
@@ -0,0 +1,200 @@
+static_assert(AssertSize<CardIdentificationRegister, 16>());
+
+union SDConfigurationRegister {
+    u8 raw[8];
```
Is there a reason this is not a `u64`?

mrkct replied:
I'm passing it to `UserOrKernelBuffer::for_kernel_buffer` (which takes a `u8*`) in `retrieve_sd_configuration_register` ( SDHostController.cpp )

Hendiadyoin1 commented on Kernel/Storage/SD/Registers.h:
```diff
@@ -0,0 +1,200 @@
+
+// CID response: "20 bits of the response data (R[127:8]) stored in the Response register at REP[119:0]"
+union CardIdentificationRegister {
+    u32 raw[4];
+
+    struct [[gnu::packed]] {
+        u32 manufacturing_date : 12;
+        u32 : 4;
+        u32 product_serial_number : 32;
+        u32 product_revision : 8;
+        u64 product_name : 40;
+        u32 oem_id : 16;
+        u32 manufacturer_id : 8;
+    };
```
See above, maybe use a u64 for everything here

Hendiadyoin1 replied:
Edit: I see why its a `u32 raw[4]`, because thats what `Response has aswell`
But for the struct (which does not need to be packed) please use `u64` uniformly

supercomputer7 commented on Kernel/Storage/SD/Registers.h:
```diff
@@ -0,0 +1,200 @@
+        u32 sd_security : 3;
+        u32 sd_bus_widths : 4;
+        u32 sd_specification3 : 1;
+        u32 extended_security : 4;
```
I really don't like this. Bit-fields are fine for data you want to pack but for MMIO access this is certainly not acceptable, especially when we do want to be compatible with big endian CPUs (although we support non currently, it's a good practice).

supercomputer7 replied:
Please use bit-wise operations as we use those all over the kernel codebase instead.

Hendiadyoin1 replied:
As far as I can tell this seems to be a struct locate in MMIO, but copied out of and then accessed

Hendiadyoin1 replied:
If the enidanness becomes a problem,
GCC seems to have `__attribute__ ((scalar_storage_order("little-endian")))`, not sure about clang

Hendiadyoin1 commented on Kernel/Storage/SD/SDHostController.cpp:
```diff
@@ -0,0 +1,667 @@
+{
+#if ARCH(AARCH64)
+    auto hardware_relative_controller_id = StorageManagement::generate_relative_sd_controller_id({});
+    auto controller = TRY(adopt_nonnull_ref_or_enomem(new RPi::SDHostController(hardware_relative_controller_id)));
```
I think this should use `new (nothrow) T`, to actually allow new to return a nullptr

Hendiadyoin1 commented on Kernel/Storage/SD/SDHostController.cpp:
```diff
@@ -0,0 +1,667 @@
+    // 7. Send CMD2 (ALL_SEND_CID) to the card
+    TRY(issue_command(SD::CommandIndex::AllSendCid, 0));
+    auto all_send_cid_response = TRY(wait_for_response());
+    SD::CardIdentificationRegister cid = { .raw = {
```
We can just `bit_cast` here I think

Hendiadyoin1 replied:
And in some other places aswell

Hendiadyoin1 commented on Kernel/Storage/SD/SDHostController.cpp:
```diff
@@ -0,0 +1,667 @@
+    // PLSS: 5.3 CSD Register
+    TRY(issue_command(SD::CommandIndex::SendCsd, rca));
+    auto send_csd_response = TRY(wait_for_response());
+    SD::CardSpecificDataRegister csd = { .raw = {
```
same

Hendiadyoin1 commented on Kernel/Storage/SD/SDHostController.cpp:
```diff
@@ -0,0 +1,667 @@
+    case SD::ResponseType::NoResponse:
+        break;
+    case SD::ResponseType::ResponseOf136Bits:
+        r.response[0] = m_registers->response_0;
+        r.response[1] = m_registers->response_1;
+        r.response[2] = m_registers->response_2;
+        r.response[3] = m_registers->response_3;
+        break;
+    case SD::ResponseType::ResponseOf48Bits:
+        r.response[0] = m_registers->response_0;
```
Correct me if I am wrong, but the bit numbers dont seem to add up
in the first case we read 128 out of 136 (Delta of 8 bit) and in the second 32 out of 48 (Delta of 16 bit)

Maybe add a omment to explain the discrepancies

mrkct replied:
That's because there are technically some extra bits in the physical communication layer, stuff like start&end bits and a checksum. We don't receive any of that stuff at our level of abstraction though.

I have added a comment explaining this.

Hendiadyoin1 commented on Kernel/Storage/SD/SDHostController.cpp:
```diff
@@ -0,0 +1,667 @@
+        // Interrupt has occurred, go to step (6).
+
+        while ((m_registers->interrupt_status & transfer_complete) == 0)
+            ;
```
please call either one of `Processor::pause` or `Processor::wait_check` or `Processor::wait_for_interrupt`, unless you want the pi to catch fire

mrkct replied:
I decided to remove this IF-branch and instead slap a TODO on it.

It's never reached and I'm not sure this function is fit to handle that type of commands anyway, I'd rather have someone fall into a TODO and find the issue quickly

Hendiadyoin1 commented on Kernel/Storage/SD/SDHostController.cpp:
```diff
@@ -0,0 +1,667 @@
+
+    // FIXME: The way the SD Clock is to be calculated is different for other
+    // versions
+    VERIFY(host_version() == SD::HostVersion::Version3);
```
```suggestion
    if(host_version() != SD::HostVersion::Version3)
        TODO();
```

Hendiadyoin1 commented on Kernel/Storage/SD/SDHostController.cpp:
```diff
@@ -0,0 +1,667 @@
+    // FIXME: The way the SD Clock is to be calculated is different for other
+    // versions
+    VERIFY(host_version() == SD::HostVersion::Version3);
+    u32 divisor = AK::max(sd_clock_frequency / (frequency), 2);
+
+    if (sd_clock_frequency / divisor >= frequency) {
+        divisor++;
+    }
```
could this be achieved with a `ceil_div`?

Hendiadyoin1 replied:
(If I read that algorithm correctly)

mrkct replied:
Yes

Hendiadyoin1 commented on Kernel/Storage/SD/SDHostController.cpp:
```diff
@@ -0,0 +1,667 @@
+
+    // 2. Set **Internal Clock Enable** and **SDCLK Frequency Select** in the
+    // *Clock Control* register
+    const u32 two_upper_bits_of_sdclk_frequency_select = (divisor >> 8 & 0x3)
+        << 6;
+    const u32 eight_lower_bits_of_sdclk_frequency_select = (divisor & 0xff)
+        << 8;
```
wierd line breaks

Hendiadyoin1 commented on Kernel/Storage/SD/SDHostController.cpp:
```diff
@@ -0,0 +1,667 @@
+    const u32 eight_lower_bits_of_sdclk_frequency_select = (divisor & 0xff)
+        << 8;
+    const u32 sdclk_frequency_select = two_upper_bits_of_sdclk_frequency_select | eight_lower_bits_of_sdclk_frequency_select;
+    m_registers->host_configuration_1 = m_registers->host_configuration_1 | internal_clock_enable | sdclk_frequency_select;
```
```suggestion
    m_registers->host_configuration_1 |= internal_clock_enable | sdclk_frequency_select;
```

mrkct replied:
Can't do this because `m_registers` is volatile. It's a compile error:
```error: compound assignment with 'volatile'-qualified left operand is deprecated [-Werror=volatile]```

Hendiadyoin1 replied:
OK did not know this was an error

Hendiadyoin1 commented on Kernel/Storage/SD/SDHostController.cpp:
```diff
@@ -0,0 +1,667 @@
+    }
+
+    // 4. Set **SD Clock Enable** in the *Clock Control* register to 1
+    m_registers->host_configuration_1 = m_registers->host_configuration_1 | sd_clock_enable;
```
```suggestion
    m_registers->host_configuration_1 |=  sd_clock_enable;

Hendiadyoin1 commented on Kernel/Storage/SD/SDHostController.cpp:
```diff
@@ -0,0 +1,667 @@
+    // 3.2.2 SD Clock Stop Sequence
+
+    // 1. Set **SD Clock Enable** in the *Clock Control* register to 0
+    m_registers->host_configuration_1 = m_registers->host_configuration_1 & ~sd_clock_enable;
```
see above

Hendiadyoin1 commented on Kernel/Storage/SD/SDHostController.cpp:
```diff
@@ -0,0 +1,667 @@
+ErrorOr<void> SDHostController::reset_host_controller()
+{
+    m_registers->host_configuration_0 = 0;
+    m_registers->host_configuration_1 = m_registers->host_configuration_1 | software_reset_for_all;
```
see above

Hendiadyoin1 commented on Kernel/Storage/SD/SDHostController.cpp:
```diff
@@ -0,0 +1,667 @@
+    // block to Block Size register.
+    // 2. Set the value corresponding to the executed data block count to Block
+    // Count register in accordance with Table 2-8.
+    m_registers->block_size_and_block_count = (block_count << 16) | block_size;
```
why arent these 2 fields instead?

mrkct replied:
Not sure how relevant this is (maybe the compiler already handles this) but the spec for the RPi says that ["the EMMC module registers can only be accessed as 32 bit registers"](https://pdf1.alldatasheet.com/pdfjsview/web/viewer.html?file=//pdf1.alldatasheet.com/datasheet-pdf/view/502533/BOARDCOM/BCM2835/+_035W5VvMzS..SReHIuCHEh+/datasheet.pdf) (page 66, first paragraph of "Registers").
I wanted to be explicit when writing to these registers.

Also, in case, should I still change this after what @supercomputer7 said above?

Hendiadyoin1 replied:
If this is directly read from MMIO then you definitely should leave it 32 bit, the `volatile` should restrict it to only do full 32 bit reads then, IIRC

Hendiadyoin1 commented on Kernel/Storage/SD/SDHostController.cpp:
```diff
@@ -0,0 +1,667 @@
+    bool contains_errors = false;
+    switch (index) {
+    case SD::CommandIndex::SendRelativeAddr:
+        if (status.error || status.illegal_command || status.com_crc_error) {
```
you dont necessarily need to use curly braces for single line bodies

Hendiadyoin1 reviewed (commented):
A bit more from me

And can we get atleast one spec link?

konradekk commented on Kernel/Storage/SD/Commands.cpp:
```diff
@@ -0,0 +1,194 @@
+    cmd.response_type = ResponseType::ResponseOf48Bits;
+    cmd.is_data = true;
+    cmd.crc_enable = true;
+    // 1.11 Auto CMD12
+    // The Host Driver should set Auto CMD12 Enable in the Transfer
+    // Mode register when issuing a multiple block transfer command
+    cmd.auto_command = SendAutoCommand::Command12;
+    cmd.block_counter = true;
+    cmd.multiblock = true;
```
(is this something extractable?)

mrkct replied:
There are only two commands that set those bits this way, feels overkill to extract a function just for that.

konradekk commented on Kernel/Storage/SD/Commands.cpp:
```diff
@@ -0,0 +1,194 @@
+
+namespace Kernel::SD {
+
+constexpr Command build_cmd0()
```
itʼ be nice to describe the comments a little (provide a human-readable name?). pointing to the specs would also be a treat! 🙇🏻‍♂️

(i wouldnʼt mind using command index enum names instead of using `cmd✖️` pattern — that could go to the comment instead? 🤔)

mrkct replied:
I can link the "Physical Layer" spec, that one has a big table with all commands, a description of what they do and what they return + their command index. It doesn't tell you how to set all the bits, but you can sort of figure it out from there.

I don't think there's a place in the spec explaining what to set for those bit, but maybe I simply didn't find it. @Ollrogge wrote this file, maybe you have more info?

> (i wouldnʼt mind using command index enum names instead of using cmd✖️ pattern — that could go to the comment instead? thinking)

I don't know, these are only used to fill the array in `get_command`. Having the command number there makes it easier to see where there are gaps

Ollrogge replied:
No there is no place where the spec explains how to set the bits. I did it the way @mrkct expained it. Looked at what the commands return, how they are used and at other implementations of the SDHC spec and tried to figure it out this way.

Having the command numbers there also makes it easier to find them in the spec because it only talks about e.g. "CMD0" and not about the "GoIdleState command"

konradekk commented on Kernel/Storage/SD/Registers.h:
```diff
@@ -0,0 +1,200 @@
+    u32 raw;
+    struct {
+        u32 : 15;
+        u32 vdd_voltage_window_27_28 : 1;
```
boolean for `: 1` could be easier to access… (not sure how relevant is this exactly…)

Hendiadyoin1 replied:
I disagree
see <https://github.com/SerenityOS/serenity/pull/17887#discussion_r1140880456>

konradekk commented on Kernel/Storage/StorageManagement.cpp:
```diff
@@ -445,14 +439,11 @@ UNMAP_AFTER_INIT void StorageManagement::initialize(StringView root_device, bool
-    } else {
-        m_controllers.append(controller.release_value());
+
+    if (auto sdhc = SDHostController::try_initialize(); !sdhc.is_error()) {
```
maybe debug-print an error?

mrkct replied:
Good idea. Decided to print an error via `dmesgl` if `try_initialize` returns an error and it's not because there's no sd card slot

Hendiadyoin1 commented on Kernel/Storage/SD/Commands.h:
```diff
@@ -0,0 +1,97 @@
+
+    struct {
+        u32 : 1;
+        bool block_counter : 1;
```
I think these should be `u32` as well

Hendiadyoin1 commented on Kernel/Storage/SD/SDHostController.cpp:
```diff
@@ -0,0 +1,623 @@
+    auto start = TimeManagement::the().monotonic_time().to_nanoseconds();
+    auto end = start + nanoseconds;
+    while (TimeManagement::the().monotonic_time().to_nanoseconds() < end)
+        ;
```
please use a hot loop hint here
(`Processor::pause` or `Processor::wait_check`)

Hendiadyoin1 commented on Kernel/Storage/SD/SDHostController.cpp:
```diff
@@ -0,0 +1,623 @@
+    TRY(issue_command(SD::CommandIndex::AppSetBusWidth, 0x2)); // 0b00=1 bit bus, 0b10=4 bit bus
+    TRY(wait_for_response());
+
+    return TRY(adopt_nonnull_lock_ref_or_enomem(
+        new SDMemoryCard(*this,
```
again only `new (nothrow)` can actually return a nullptr signaling ENOMEM, so prefer that

konradekk reviewed (approved):
🤯

FireFox317 commented on Kernel/Storage/StorageManagement.cpp:
```diff
@@ -67,6 +69,12 @@ u32 StorageManagement::generate_relative_ata_controller_id(Badge<ATAController>)
     s_relative_ata_controller_id++;
     return controller_id;
 }
+u32 StorageManagement::generate_relative_sd_controller_id(Badge<SDHostController>)
```
Missing empty line.

FireFox317 commented on Kernel/Storage/SD/SDMemoryCard.h:
```diff
@@ -0,0 +1,49 @@
+private:
+    SDMemoryCard(SDHostController& sdhc, StorageDevice::LUNAddress, u32 hardware_relative_controller_id, u32 block_len, u64 capacity_in_blocks, u32 relative_card_address, SD::OperatingConditionRegister ocr, SD::CardIdentificationRegister cid, SD::SDConfigurationRegister scr);
+
+    enum class CardAddressingMode { ByteAddressing,
+        BlockAddressing };
```
Formatting seems a bit weird here, I would have expected something like:
```c++
enum class CardAddressingMode {
    ByteAddressing,
    BlockAddressing,
};
```
This seems to be how we normally format these enums.

FireFox317 reviewed (approved):
Just some tiny style issues, but this is really nice!

I also tested this, and it works very good! I think this is ready to be merged.

Thanks for working on this!

mrkct commented:
After porting this driver to x86 too, I realized I had made some small mistakes here and there which I have fixed.
Nothing really big, but just to be clear:
- I have changed `SDHostController`'s constructor to take zero arguments. It previously took a relative hardware id as an argument, but that made no sense since the id must be unique and it's always generated by `generate_relative_sd_controller_id`  
- I'm now registering the SD card storage device with DeviceManagement. Not doing it was an error that would have shown up later in the boot process probably  
- Made `SDHostController::initialize` a public method, and removed the old static `try_initialize`, which contained hard coded logic for the RPi's controller and simply created an instance of the RPi's controller and called `initialize` on it
- Turned the RPi's controller into a singleton. 
- Removed useless `friend class`
